NitroxServer/Serialization/SaveDataUpgrades/SaveDataUpgrade.cs

```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.IO;
+
+namespace NitroxServer.Serialization.Upgrade
+{
+    public abstract class SaveDataUpgrade
+    {
+        public abstract Version TargetVersion { get; }
+
+        public void UpgradeData(string saveDir, string fileEnding)
+        {
+            string baseDataPath = Path.Combine(saveDir, "BaseData" + fileEnding);
+            string playerDataPath = Path.Combine(saveDir, "PlayerData" + fileEnding);
+            string worldDataPath = Path.Combine(saveDir, "WorldData" + fileEnding);
+            string entityDataPath = Path.Combine(saveDir, "EntityData" + fileEnding);
+
+            File.WriteAllText(baseDataPath, UpgradeBaseData(File.ReadAllText(baseDataPath)));
+            File.WriteAllText(playerDataPath, UpgradePlayerData(File.ReadAllText(playerDataPath)));
+            File.WriteAllText(worldDataPath, UpgradeWorldData(File.ReadAllText(worldDataPath)));
+            File.WriteAllText(entityDataPath, UpgradeEntityData(File.ReadAllText(entityDataPath)));
+        }
+
+        protected virtual string UpgradeBaseData(string data)
+        {
+            return data;
+        }
```

Comment:
Wouldn't it be better to version the data models themselves instead of mutating raw string data?  IMO, upgrade base data should take a BaseData_1_3 and output a BaseData_1_4.

Reply:
Yeah that would be the better approach but then we would need to save each iteration of the save file structure with all changing classes which would probably be too much after a handful of version.

Reply:
I mean we can remove those changes if the version is 5 versions prior or more to help keep bloat down

Reply:
I counted and we have around 25 files/classes in NitroxServer used for saving, not including gamelogic stored in NitroxModel. Even if we just support 5 versions it will bloat up the program and the exported dll.

Reply:
To be clear, I only think we should master a new version when the model changes, not a duplicate of every model for every version.  However, we should likely only start that when in beta as the changes right now still aren't backwards compatible even if you update the model (missing stuff in the world).  However, we can define the framework for it now.

Reply:
Yeah, editing raw json is not the ideal way but as you said it lays a foundation (like saving the assembly version in the save files). I will look into making the json editing easier/more useful. 

The current solution should just be for things easy to fix and something like "adding new values to models" will break it, but if we just remove a field it is fixable by this early approach.

Reply:
I switched from string based editing to `JObject` which should improve editing a lot.